# <font color='red'><b>Bootstrap Model with Random Forest</b> </font>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [3]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [4]:
x.shape

(506, 13)

In [5]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

In [7]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples
    selected_rows = np.random.choice(input_data.shape[0], 303, replace=False)
    r_rows = np.random.choice(selected_rows.shape[0], 203, replace=True)
    selected_columns = np.random.choice(input_data.shape[1], np.random.randint(3,input_data.shape[1],1), replace=False)
    sample_data = input_data[selected_rows[:,None], selected_columns]
    target_data = target_data[selected_rows]
    
    # replacting_data
    replicated_sample = sample_data[r_rows]
    target_rep = target_data[r_rows]
    
    # Concatenating data
    sampled_input_data = np.vstack((sample_data, replicated_sample))
    sample_target_data = np.concatenate((target_data, target_rep), axis=0)
    

    return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists

*  <font color='blue'> <b>Create 30 samples </b>

In [9]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='red'><b>Step - 2 </b></font>

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [12]:
from sklearn.ensemble import RandomForestRegressor
trees = []
for i in range(30):
    dt = RandomForestRegressor(max_depth=None)
    input_data = list_input_data[i]
    target_data = list_output_data[i]
    dt.fit(input_data, target_data)
    trees.append(dt)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [26]:
predictions = []
for i in range(30):
    predictions.append(trees[i].predict(list_input_data[i]))
    

mse_train = []
for i in range(30):
    y = list_output_data[i]
    mse_train.append(mean_squared_error(predictions[i], y))

for i in range(30):
    print("MSE value of sample{} is {}".format(i+1, mse_train[i]))    

MSE value of sample1 is 44.475134832947276
MSE value of sample2 is 52.245032672402324
MSE value of sample3 is 43.16755897002427
MSE value of sample4 is 55.03879187436881
MSE value of sample5 is 44.267905638362045
MSE value of sample6 is 44.53910589293804
MSE value of sample7 is 47.19360078456453
MSE value of sample8 is 46.590184620758755
MSE value of sample9 is 47.78147318337792
MSE value of sample10 is 47.46473254349309
MSE value of sample11 is 47.6514310934195
MSE value of sample12 is 51.85512214899914
MSE value of sample13 is 47.62382216181704
MSE value of sample14 is 46.006847196400564
MSE value of sample15 is 43.89358448023576
MSE value of sample16 is 47.33969529510235
MSE value of sample17 is 49.157856992148666
MSE value of sample18 is 45.85830431477411
MSE value of sample19 is 47.798062634571785
MSE value of sample20 is 49.013359245934645
MSE value of sample21 is 44.974854383003105
MSE value of sample22 is 50.92671199404856
MSE value of sample23 is 47.551010158696634
MSE value o

In [27]:
boston = load_boston()
x = boston.data #independent variables
y=boston.target #target variable
pred_cor = list()
for i in range(30):
    pred_cor.append(trees[i].predict(x[:,list_selected_columns[i]]))

pred_y = np.zeros(x.shape[0])
for i in range(x.shape[0]):
    for j in range(30):
        pred_y[i] += pred_cor[j][i]
    pred_y[i] /= 30

mse_pop = mean_squared_error(y, pred_y)
print(mse_pop)

83.98216540356778


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [14]:
pred_cor = list()
for i in range(30):
    pred_cor.append(trees[i].predict(x[:,list_selected_columns[i]]))

pred_y_oob = np.zeros(x.shape[0])
for i in (range(x.shape[0])):
    count_b = 0
    for j in range(30):
        count = 0
        for k in range(x.shape[0]):
            if np.all(np.equal(x[i,list_selected_columns[j]],np.array(list_input_data[j][k,0:len(list_input_data[j][0])]))):
                count += 1
                break
        if count != 0:
            continue
        pred_y_oob[i] += pred_cor[j][i]
        count_b += 1
    pred_y_oob[i] /= count_b

mse_oob_pop = mean_squared_error(y,pred_y_oob)
print(mse_oob_pop)

95.90761661295751


In [15]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples
    selected_rows = np.random.choice(input_data.shape[0], 303, replace=False)
    r_rows = np.random.choice(selected_rows.shape[0], 203, replace=True)
    selected_columns = np.random.choice(input_data.shape[1], np.random.randint(3,input_data.shape[1],1), replace=False)
    sample_data = input_data[selected_rows[:,None], selected_columns]
    target_data = target_data[selected_rows]
    
    # replacting_data
    replicated_sample = sample_data[r_rows]
    target_rep = target_data[r_rows]
    
    # Concatenating data
    sampled_input_data = np.vstack((sample_data, replicated_sample))
    sample_target_data = np.concatenate((target_data, target_rep), axis=0)
    

    return sampled_input_data , sampled_target_data,selected_rows,selected_columns


In [16]:
from tqdm import tqdm
boston = load_boston()
x = boston.data #independent variables
y=boston.target #target variable
   
    
mse_corpus = list()
mse_oob = list()
for _ in tqdm(range(35)):
        
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]
    for i in range(30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)

        
    trees = list() #training the 30 DTs
    for i in range(30):
        dt = RandomForestRegressor(max_depth=None)
        input_data = list_input_data[i]
        target_data = list_output_data[i]
        dt.fit(input_data, target_data)
        trees.append(dt)
        
    x = boston.data
    pred_cor = list()
    for i in range(30):
        pred_cor.append(trees[i].predict(x[:,list_selected_columns[i]]))
        
    y = boston.target
    pred_y = np.zeros(x.shape[0])
    for i in range(x.shape[0]):
        for j in range(30):
            pred_y[i] += pred_cor[j][i]
        pred_y[i] /= 30
            
    m = mean_squared_error(y,pred_y)
    mse_corpus.append(m)
        
    pred_y_oob = np.zeros(x.shape[0])
    for i in (range(x.shape[0])):
        count_b = 0
        for j in range(30):
            count = 0
            for k in range(x.shape[0]):
                if np.all(np.equal(x[i,list_selected_columns[j]],np.array(list_input_data[j][k,0:len(list_input_data[j][0])]))):
                    count += 1
                    break
            if count != 0:
                continue
            pred_y_oob[i] += pred_cor[j][i]
            count_b += 1
        pred_y_oob[i] /= count_b
            
    ms = mean_squared_error(y, pred_y_oob)
    mse_oob.append(ms)

m = np.asarray(mse_corpus)
mo = np.asarray(mse_oob)        
        

100%|██████████████████████████████████████████████████████████████████████████████████| 35/35 [35:27<00:00, 60.80s/it]


In [25]:
print(m)
print(mo)

[88.71937163 86.9335312  85.69196418 87.56384463 86.43665058 85.2727732
 87.36122717 86.75667714 86.78784384 86.31381278 83.98323058 86.17012706
 86.55096764 84.77892916 86.23984195 86.74507753 88.35502449 84.39642928
 86.87015105 84.73242319 86.43920158 86.4680041  85.7563917  86.66632432
 84.1716166  86.06753496 86.48279509 83.65789352 84.6232441  84.63344815
 86.60863625 86.55909744 90.36418039 86.41763521 83.9821654 ]
[89.57488665 88.66683496 85.8959892  87.96952901 86.51611953 85.34093686
 87.67729504 88.68527316 88.14979257 88.44437705 85.21488608 87.11740276
 87.08021256 84.6778408  87.23845936 86.48876928 88.80838944 83.61214892
 87.90172128 86.04858511 85.73831531 87.16603976 85.34019089 87.07898666
 84.34197151 87.36338213 87.98005261 85.24595823 84.49388999 84.91827689
 88.28318615 86.51549965 91.75129002 87.69037764 83.08345681]


In [20]:
from prettytable import PrettyTable
import random



x = PrettyTable()
x = PrettyTable(["#samples", "Sample Size", "Sample mean", "Population Mean","Sample Std","Left C.I","Right C.I", "Catch"])
for i in range(10):
    sample=m[random.sample(range(0, m.shape[0]), 20)]
    sample_mean = sample.mean()
    sample_std =  sample.std()
    sample_size = len(sample)
    # here we are using sample standard deviation instead of population standard deviation
    left_limit  = np.round(sample_mean - 2*(sample_std/np.sqrt(sample_size)), 3)
    right_limit = np.round(sample_mean + 2*(sample_std/np.sqrt(sample_size)), 3)
    row = []
    row.append(i+1)
    row.append(sample_size)
    row.append(sample_mean)
    row.append(mse_pop)
    row.append(np.round(sample_std/np.sqrt(sample_size),3))
    row.append(left_limit)
    row.append(right_limit)
    row.append((mse_pop <= right_limit) and (mse_pop >= left_limit))
    x.add_row(row)
print(x)
print("C.I of mse_scores")


+----------+-------------+-------------------+-------------------+------------+----------+-----------+-------+
| #samples | Sample Size |    Sample mean    |  Population Mean  | Sample Std | Left C.I | Right C.I | Catch |
+----------+-------------+-------------------+-------------------+------------+----------+-----------+-------+
|    1     |      20     | 85.99108142272061 | 83.98216540356778 |   0.334    |  85.324  |   86.658  | False |
|    2     |      20     | 86.09939897178191 | 83.98216540356778 |   0.194    |  85.712  |   86.487  | False |
|    3     |      20     | 86.52760910060496 | 83.98216540356778 |   0.339    |  85.849  |   87.206  | False |
|    4     |      20     | 86.24941122402782 | 83.98216540356778 |    0.32    |  85.61   |   86.889  | False |
|    5     |      20     | 86.41983980600041 | 83.98216540356778 |   0.319    |  85.782  |   87.057  | False |
|    6     |      20     | 85.84709953432154 | 83.98216540356778 |   0.314    |  85.219  |   86.476  | False |
|

In [21]:
x = PrettyTable()
x = PrettyTable(["#samples", "Sample Size", "Sample mean", "Population Mean","Sample Std","Left C.I","Right C.I", "Catch"])
for i in range(10):
    sample=mo[random.sample(range(0, mo.shape[0]), 20)]
    sample_mean = sample.mean()
    sample_std =  sample.std()
    sample_size = len(sample)
    # here we are using sample standard deviation instead of population standard deviation
    left_limit  = np.round(sample_mean - 2*(sample_std/np.sqrt(sample_size)), 3)
    right_limit = np.round(sample_mean + 2*(sample_std/np.sqrt(sample_size)), 3)
    row = []
    row.append(i+1)
    row.append(sample_size)
    row.append(sample_mean)
    row.append(mse_oob_pop)
    row.append(np.round(sample_std/np.sqrt(sample_size),3))
    row.append(left_limit)
    row.append(right_limit)
    row.append((mse_oob_pop <= right_limit) and (mse_oob_pop >= left_limit))
    x.add_row(row)
print(x)
print("C.I of oob mse_scores")

+----------+-------------+-------------------+-------------------+------------+----------+-----------+-------+
| #samples | Sample Size |    Sample mean    |  Population Mean  | Sample Std | Left C.I | Right C.I | Catch |
+----------+-------------+-------------------+-------------------+------------+----------+-----------+-------+
|    1     |      20     | 86.85022995429458 | 95.90761661295751 |   0.366    |  86.118  |   87.583  | False |
|    2     |      20     | 86.46056641959248 | 95.90761661295751 |    0.37    |  85.72   |   87.201  | False |
|    3     |      20     | 86.78841685390535 | 95.90761661295751 |   0.363    |  86.062  |   87.515  | False |
|    4     |      20     | 86.77612163276436 | 95.90761661295751 |   0.286    |  86.204  |   87.349  | False |
|    5     |      20     | 86.71275931994988 | 95.90761661295751 |   0.467    |  85.779  |   87.646  | False |
|    6     |      20     | 86.93952438409671 | 95.90761661295751 |   0.432    |  86.076  |   87.803  | False |
|

 We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

In [23]:
xq = [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
xq = np.asarray(xq)
pred_cor = list()
for i in range(30):
    bh = xq[list_selected_columns[i]].reshape(1,-1)
    pred_cor.append(trees[i].predict(bh))

In [24]:
pred_y = 0
for j in range(30):
    pred_y += pred_cor[j]
pred_y /= 30

print("Predicted value of xq is {}".format(pred_y))

Predicted value of xq is [25.37375794]


<br><br><br>